retranmission.go

```diff
@@ -0,0 +1,137 @@
+package main
+
+import (
+	"github.com/go-errors/errors"
+	"github.com/lightningnetwork/lnd/lnwire"
+)
+
+// MessageStore represents the retransmission subsystem storage for lnwire
+// messages, it might be boltd storage, local storage, test storage or hybrid
+// one.
+//
+// NOTE: The original purpose of creating this interface was in separation
+// between the retransmission logic and specifics of storage itself. In case of
+// such interface we may create the test storage and register the add,remove,get
+// actions without the need of population of lnwire messages with data.
+type MessageStore interface {
+	// Get returns the sorted set of messages in the order they were
+	// added in storage.
+	Get() ([]lnwire.Message, error)
+
+	// Add adds new message in the storage with preserving the order.
+	Add(msg lnwire.Message) error
+
+	// Remove deletes all messages that corresponds to codes from storage.
+	//
+	// Note: Why this function gets set of codes as input?
+	// Is due to the fact of logic of retransmission subsystem, where we
+	// need remove messages not one by one but in contrary by groups of
+	// messages.
+	Remove(codes ...lnwire.Code) error
+}
+
+// retransmitter represents the retransmission subsystem which is described
+// in details in BOLT #2 (Message Retransmission). This subsystem keeps
+// records of all messages that were sent to other peer and waits the ACK
+// message to be received from other side. The ACK message denotes that the
+// previous messages was read. Because communication transports are unreliable
+// and may need to be re-established from time to time and reconnection
+// introduces doubt as to what has been received such logic is needed to be sure
+// that peers are in consistent state in terms of message communication.
+type retransmitter struct {
+	// storage is a message storage which is needed to store the messages
+	// which weren't acked and restore them after node restart, in order to
+	// send them to other side again.
+	storage MessageStore
+}
+
+// newRetransmitter creates new instance of retransmitter.
+func newRetransmitter(storage MessageStore) (*retransmitter, error) {
+	return &retransmitter{storage: storage}, nil
+}
+
+// Register adds message that should be acknowledged in the message storage.
+func (rt *retransmitter) Register(msg lnwire.Message) error {
+	switch msg.Command() {
+	// messages without acknowledgment
+	case lnwire.CmdCloseComplete,
+		lnwire.CmdChannelUpdateAnnouncement,
+		lnwire.CmdChannelAnnouncement,
+		lnwire.CmdNodeAnnouncement,
+		lnwire.CmdPing,
+		lnwire.CmdPong,
+		lnwire.CmdErrorGeneric,
+		lnwire.CmdInit:
+		return nil
+	default:
+		return rt.storage.Add(msg)
+	}
+}
+
+// Ack encapsulates the specification logic about which messages should be
+// acknowledged by receiving this one.
+func (rt *retransmitter) Ack(msg lnwire.Message) error {
```

Comment:
With the comment above, this message would be simplified a good bit as it would attempt to perform a single unconditional delete from the database. 

The mapping here (what get's deleted on receipt of a message) would be moved into the storage layer as it would need to be consulted each time a message is written.